base/src/commonMain/kotlin/flow/BaseFlowable.kt

```diff
@@ -18,25 +18,35 @@ Copyright 2019 Splendo Consulting B.V. The Netherlands
 */
 
 import com.splendo.kaluga.base.runBlocking
+import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.channels.BroadcastChannel
 import kotlinx.coroutines.channels.ConflatedBroadcastChannel
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.asFlow
 
-open class BaseFlowable<T>(private val channel: BroadcastChannel<T> = ConflatedBroadcastChannel()) : Flowable<T> {
+/**
+ * Base definition of [Flowable]. Abstract class, use [ColdFlowable] or [HotFlowable] instead.
+ *
+ * @param T the value type to flow on.
+ * @param channelFactory Factory for generating a [BroadcastChannel] on which the data is flown
+ */
+abstract class BaseFlowable<T>(private val channelFactory: () -> BroadcastChannel<T> = { ConflatedBroadcastChannel() }) : Flowable<T> {
 
-    final override fun flow(flowConfig: FlowConfig): Flow<T> {
-        return flowConfig.apply(channel.asFlow())
+    protected var channel = lazy {channelFactory() }
```

Comment:
So I did some refactoring here.

From now on all BaseFlowable can be instantly closed by calling `close`. This closes the underlying channel, and a a result all observing FlowCollectors. In addition, ColdFlowable has been given an initialization property that allows it to trigger auto closing when the number of observers decreases to 0. This is set to `true` by default.

To ensure initialization behaviour, Flowable only allows data to be set once the underlying channel has been created (immediately for HotFlowable, after the first observer on ColdFlowable)

Reply:
Ive also made ColdStateRepo create new ColdFlowable when observation is stopped and then resumed. This is important to keep because you dont want to recreate StateRepos every time the observers stop.

Reply:
Looks good.